### PR TITLE
feat: added reason column to federation_blocklist

### PR DESCRIPTION
**Pull request by TKilFree**
This allows blocked instances to (optionally) have a reason set. 

Addresses issue: https://github.com/LemmyNet/lemmy/issues/1947

Linked UI MR: https://github.com/LemmyNet/lemmy-ui/pull/1367

I've just picked this issue up (by searching for 'good first issue's) as a way of dipping my feet into this project and I've only really tinkered with rust in toy projects before so please let me know if there's any problems.

### CODE REVIEW

File: migrations/2023-06-17-132316_add_reason_to_blocklist/up.sql
Comment (dessalines) [inline migrations/2023-06-17-132316_add_reason_to_blocklist/up.sql:4]: You don't need null here.

File: migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql
Comment (dessalines) [inline migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql:3]: This and the one below it, have it be a single line.

File: migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql
Comment (dessalines) [inline migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql:1]: I'm somewhat torn, because nearly every `reason` we have is in the `mod_` tables... the idea being that each action can leave a trace in the modlog. Without modlog tables, those changes in reason, as well as blocking / unblocking instances could get lost.

But I think for this one, its fine, since it'd also be nice to show the reason on the instances page, without having to do any weird sql joins grabbing the most recent reason.

File: crates/api_common/src/site.rs
Comment (dessalines) [inline crates/api_common/src/site.rs:265]: This definitely needs to be reworked... Rather than having a separate map just for reasons. `blocked_instances: Option<Vec<BlockAction..`, where BlockAction contains the optional reason along with the instance domain string.

It will probably be a while on this then, as I'd want to get @SleeplessOne1917 's input.

File: crates/db_schema/src/impls/federation_blocklist.rs
Comment (dessalines) [inline crates/db_schema/src/impls/federation_blocklist.rs:16]: This needs to be a single list, not a list and a map.

File: crates/api_common/src/site.rs
Comment (TKilFree) [reply crates/api_common/src/site.rs:265]: definitely agree in principle , this was me trying to not make a breaking API change as i'm assuming there's a whole versioning process for breaking changes?

File: migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql
Comment (TKilFree) [reply migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql:1]: what would ur opinion be on having corresponding mod tables for changes to allow/blocklists, with the source of truth for current values still being the `federation_blocklist`/`federation_allowlist` tables? effectively the mod tables would be audit logs rather than something the application itself would refer to

File: crates/db_schema/src/impls/federation_blocklist.rs
Comment (TKilFree) [reply crates/db_schema/src/impls/federation_blocklist.rs:16]: 👍  this should happen as a consequence the above

File: migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql
Comment (dessalines) [reply migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql:1]: I'd def be good with that. My issue is that I'd rather not duplicate any data, and make sure that the reason is either in one table or the other, but not both. `reason` typically goes for modlog entries, because things like bans and removes don't necessarily represent a permanent state of affairs. That probably should also apply here, where you might block, then unblock instances.

File: crates/api_common/src/site.rs
Comment (dessalines) [reply crates/api_common/src/site.rs:265]: Yep this would be a breaking API change, but considering the lemmy API hasn't really been finalized or hit a `1.0` status yet, don't be afraid to do those.

Comment (SleeplessOne1917) [reply crates/api_common/src/site.rs:265]: I don't think I'm familiar with the backend codebase or DB tables to give much useful insight on this TBH.

File: migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql
Comment (TKilFree) [reply migrations/2023-06-17-132316_add_reason_to_blocklist/down.sql:1]: i've had a quick investigation of this and while it's definitely possible, the changeset did start ballooning because of the joins (to the latest entry of) the new table. i suspect it might be better implemented as a separate commit, but at the end of the day it's up to you whether you'd rather it get done completely and correctly at this point, or whether it's ok to get this in without the log table and i can raise an issue to add it as a separate bit of work?

File: crates/db_schema/src/impls/federation_blocklist.rs
Comment (dessalines) [inline crates/db_schema/src/impls/federation_blocklist.rs:15]: Much better, thanks.

File: api_tests/src/shared.ts
Comment (dessalines) [inline api_tests/src/shared.ts:185]: I'll explain this one below.

File: crates/api_common/src/site.rs
Comment (dessalines) [reply crates/api_common/src/site.rs]: See below.

**Comment by Nutomic** [comment]
> Yes its a bit circular, but once @Nutomic approves this, I'll publish a new lemmy-js-client version with these API additions, and ping you.

These are not additions but breaking changes to existing API endpoints. So we should wait for the next major release (0.19) to merge/release this.

**Review by Nutomic — APPROVED** [review]

**Comment by Nutomic** [comment]
Looks good to me.

**Comment by TKilFree** [comment]
i've realised that if this was done without using an admin_ table now, it'd potentially be a pain to add one later (since the table would require an admin's id and therefore migrations out of the federation_blocklist table's reason column would be impossible)

just added a commit attempting to solve this using a new `admin_block_instance` table:
- this does lead to some slightly gnarly joins, but i guess we'd expect these tables to remain relatively small?
- i've not done the work to actually expose the new admin_ table to the usual modlog stuff, that seems like a chunk of work which could be done in a separate commit?
- there's some fiddly logic in `federation_blocklist::replace`. because we get a full replacement set of blocked instances, then we need to figure out which of these are new instances, which are existing instances but with a different reason and which already existing instances are no longer blocked
- i've added tests for the following scenarios:
  - going from no blocklist to some blocked instances
  - going from some blocked instances to none
  - updating the reasons on an existing blocklist
  - a backwards compat test for the bit that gets blocked instances to ensure it still works if there are blocked instances without any corresponding admin log entries

File: migrations/2023-06-27-113917_add_admin_block_instance_table/up.sql
Comment (dessalines) [inline migrations/2023-06-27-113917_add_admin_block_instance_table/up.sql:3]: Thanks, adding this as a mod table is better.

File: migrations/2023-06-27-113917_add_admin_block_instance_table/down.sql
Comment (dessalines) [inline migrations/2023-06-27-113917_add_admin_block_instance_table/down.sql:1]: Missing the table delete.

File: crates/api_common/src/site.rs
Comment (dessalines) [inline crates/api_common/src/site.rs:353]: Weird... what's this about. It should only be a modlog table now.

File: crates/api_common/src/site.rs
Comment (dessalines) [inline crates/api_common/src/site.rs:369]: No just return the Instance. The block and reasons should be included in the modlog only.

File: crates/db_schema/src/impls/federation_blocklist.rs
Comment (dessalines) [inline crates/db_schema/src/impls/federation_blocklist.rs:59]: Really weird joining, none of it is necessary. You only need to add the AdminBlockInstance view to the moderation tables.

File: crates/api_common/src/site.rs
Comment (TKilFree) [reply crates/api_common/src/site.rs:369]: this is to support the UI being able to show reasons for each blocked domain

File: crates/db_schema/src/impls/federation_blocklist.rs
Comment (TKilFree) [reply crates/db_schema/src/impls/federation_blocklist.rs:59]: i think this is necessary unless i'm misunderstanding.

the updates to the blocklist come as a complete set of blocked domains + reasons.
we've then got existing blocked domains + reasons and need to figure out what to put in the mod table, but there can be:
- newly added domains;
- existing domains with new reasons;
- existing unchanged domains;
- removed domains.

presumably we want to add entries for any newly added domains or existing domains with new reasons with a `blocked` flag since they represent admin actions. we also want to add entries for removed domains with the `blocked` flag false to represent the admin removing that domain.

i think the removed part is definitely necessary, the part figuring out created/edited could be removed, but then if you've got a list of 100 blocked domains and the admin edits the reason for one of them, or just removes one of them you'd be putting 100 unnecessary items in the admin log

**Comment by Nutomic** [comment]
Closing because its abandoned.

**Pull request closed without merging**
